### PR TITLE
(umbraCollab) Check media Parent for permissions when setting correct MediaType

### PR DESCRIPTION
**Pull request by cornehoskam**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes https://github.com/umbraco/Umbraco-CMS/issues/7735 & the underlying issue of uploading disallowed mediaTypes.

### Description
Ever since Umbraco version 8.14, the way media gets uploaded has changed. In the past, uploading media directly from a MediaPicker would respect the permissions set on the upload folder when not using the default mediaType of "Image". Ever since version 8.14 this has changed, and even if the folder does not allow the default "Image" item to be uploaded, it still forces the uploaded image to be that media type (and same with File).

With this pullrequest, as created and demonstrated during today's umbraCollab, we not just upload an image with the default 'Image' Media Type, but first try to determine if the folder we are trying to upload the item to has any specific permissions set that have the "umbracoFile" property. If that is the case, we want to use that Media Type instead of the default!

For Folders we are not able to distinguish between custom folders or the default Folder, so best practice for now would be to not allow the creation of the default Folder if the parent doesn't explicitly allow this, so that we can't create folders where they aren't allowed!

To test this Pullrequest, simply try out any mediapicker that points towards a media folder that has specific permissions set up, upload a new image, and it will upload it to the correct media type! If it does not find any items within it's permitted children with a property of alias umbracoFile, it will continue with the rest of the Umbraco logic and upload to the default type!

### CODE REVIEW

**Review by mikecp — APPROVED** [review]

**Comment by mikecp** [comment]
Approved at the end of umbracollab but needs final reviewing from @nul800sebastiaan

**Comment by cornehoskam** [comment]
@nul800sebastiaan Any updates on the review process so far? Let me know if there's something you feel is missing! 🙂

**Comment by nul800sebastiaan** [comment]
@cornehoskam and @mikecp thanks very much for the work so far! I have given it a go and noticed a few things were missing which I could've just sent you feedback about but while I was experimenting I accidentally kinda fixed them myself. One of the surprising things about https://github.com/umbraco/Umbraco-CMS/issues/7735 was that you can drag a whole folder onto the dropzone. That scenario is now also covered.

I also "fixed" when you create a folder with the `+` button in a media picker that we now return an error message instead of just not creating it. This has a drawback that the progress indicator under the folder path keeps going, but I feel like it's better to actually show a message, we can see if we can fix that last part later.

Finally, and most importantly, it seems like for whatever reason you were not rejecting files if they were going to get a mediaType not allowed under that parent. I swear I saw that working earlier, but it seemed missing so I added that aroung line 879 (`if (allowedContentTypes.Any(x => x.Alias == mediaTypeAlias) == false)`).

A video with different attempts:


https://user-images.githubusercontent.com/304656/155135633-dfa94556-1f8f-442b-ba2c-04fe479986f4.mp4


Now I have kind of a moral dilemma, technically these are breaking changes, suddenly you're no longer allowed to create folders where you were allowed to make them before. But I would argue that we're just implementing expected behavior.. The infamous "unbreaking breaking change". I'll have to talk to the team to see if this is  a problem. Personally I think it's good to make it explicit that folder creation is not allowed when it's not configured to be allowed and the previous behavior was a bug.

File: src/Umbraco.Web.BackOffice/Controllers/ContentControllerBase.cs [hunk 1]
```diff
@@ -224,7 +224,7 @@ protected void AddCancelMessage(INotificationModel display, string message)
         {
             // if there's already a default event message, don't add our default one
             IEventMessagesFactory messages = EventMessages;
-            if (messages != null && messages.GetOrDefault().GetAll().Any(x => x.IsDefaultEventMessage))
+            if (messages?.GetOrDefault() != null && messages.GetOrDefault().GetAll().Any(x => x.IsDefaultEventMessage))
```
Comment (ronaldbarendse) [inline src/Umbraco.Web.BackOffice/Controllers/ContentControllerBase.cs:227]: This is now calling `GetOrDefault()` twice, which can be avoided by using pattern matching:
```suggestion
            if (messages?.GetOrDefault() is EventMessages eventMessages &&
                eventMessages.GetAll().Any(x => x.IsDefaultEventMessage) == true)
```

Comment (ronaldbarendse) [reply src/Umbraco.Web.BackOffice/Controllers/ContentControllerBase.cs:227]: Or using multiple null-conditional operators:
```suggestion
            if (messages?.GetOrDefault()?.GetAll().Any(x => x.IsDefaultEventMessage) == true)
```

**Review by ronaldbarendse — CHANGES_REQUESTED** [review]

**Comment by mikecp** [comment]
I tend to agree with you @nul800sebastiaan, to me the previous behavior was wrong, so this should be seen as a bugfix and not a "feature".

File: src/Umbraco.Web.BackOffice/Controllers/MediaController.cs [hunk 2]
```diff
@@ -768,11 +780,51 @@ public async Task<IActionResult> PostAddFile([FromForm] string path, [FromForm]
...
+                var mediaFolderType =
+                    allMediaTypes.FirstOrDefault(x => x.Alias == mediaFolderItem.ContentType.Alias);
+
+                if (mediaFolderType != null)
+                {
+                    IMediaType mediaTypeItem = null;
+
+                    foreach (ContentTypeSort allowedContentType in mediaFolderType.AllowedContentTypes)
+                    {
+                        IMediaType checkMediaTypeItem = _mediaTypeService.Get(allowedContentType.Id.Value);
```
Comment (nzdev) [inline src/Umbraco.Web.BackOffice/Controllers/MediaController.cs:805]: Could reusue allMediaTypes here and save a call to the mediatype service

File: src/Umbraco.Web.BackOffice/Controllers/MediaController.cs [hunk 3]
```diff
@@ -768,11 +780,51 @@ public async Task<IActionResult> PostAddFile([FromForm] string path, [FromForm]
...
+                    //Only set the permission-based mediaType if we only allow 1 specific file under this parent.
+                    if (allowedContentTypes.Count() == 1 && mediaTypeItem != null)
+                    {
+                        mediaTypeAlias = mediaTypeItem.Alias;
+                    }
+                }
+            }
+            else
+            {
+                var typesAllowedAtRoot = _mediaTypeService.GetAll().Where(ct => ct.AllowedAsRoot).ToList();
```
Comment (nzdev) [inline src/Umbraco.Web.BackOffice/Controllers/MediaController.cs:824]: Could reusue allMediaTypes here and save a call to the mediatype service

File: src/Umbraco.Web.BackOffice/Controllers/MediaController.cs [hunk 4]
```diff
@@ -782,58 +834,67 @@ public async Task<IActionResult> PostAddFile([FromForm] string path, [FromForm]
 
                 if (_contentSettings.IsFileAllowedForUpload(ext))
```
Comment (nzdev) [inline src/Umbraco.Web.BackOffice/Controllers/MediaController.cs:835]: Nesting could be reduced by inverting the if statement and calling continue

File: src/Umbraco.Web.BackOffice/Controllers/MediaController.cs [hunk 5]
```diff
@@ -861,6 +922,29 @@ await using (var stream = formFile.OpenReadStream())
             return Ok(tempFiles);
         }
 
+        private bool IsFolderCreationAllowedHere(int parentId)
+        {
+            var allMediaTypes = _mediaTypeService.GetAll().ToList();
+            var isFolderAllowed = false;
+            if (parentId == Constants.System.Root)
+            {
+                var typesAllowedAtRoot = _mediaTypeService.GetAll().Where(ct => ct.AllowedAsRoot).ToList();
```
Comment (nzdev) [inline src/Umbraco.Web.BackOffice/Controllers/MediaController.cs:931]: allMediaTypes could be reused here

File: src/Umbraco.Web.BackOffice/Controllers/MediaController.cs [hunk 6]
```diff
@@ -722,10 +729,15 @@ public async Task<IActionResult> PostAddFile([FromForm] string path, [FromForm]
 
             var tempFiles = new PostedFiles();
 
-
             //in case we pass a path with a folder in it, we will create it and upload media to it.
             if (!string.IsNullOrEmpty(path))
             {
+                var isFolderAllowed = IsFolderCreationAllowedHere(parentId.Value);
+                if (isFolderAllowed == false)
```
Comment (nzdev) [inline src/Umbraco.Web.BackOffice/Controllers/MediaController.cs:736]: if (!IsFolderCreationAllowedHere(parentId.Value)) is slightly less code

File: src/Umbraco.Web.BackOffice/Controllers/MediaController.cs (hunk 2 above)
Comment (nzdev) [reply src/Umbraco.Web.BackOffice/Controllers/MediaController.cs:805]: This method is very long and will be hard to maintain. Could this part that is responsible for generating the allowed content types be made into it's own method?

**Comment by nul800sebastiaan** [comment]
Thanks @nzdev for the additional comments. Saved a bunch of calls to the `MediaTypeService`, very nice indeed! 👍 

I agree that this whole class needs a good cleanup but we'll save that for another day.

It looks like everyone was happy with the changes to make the uploader more restrictive, also known as: fixing the existing bugs to the uploader. 

So.. finally time to merge this one, thanks again for getting it started and all the work @cornehoskam!

And look at that shiny new contributor badge on your profile as well! https://our.umbraco.com/members/cornehoskam 🏅

**Comment by bachicool** [comment]
@nul800sebastiaan This is still happening in Umbraco 8.18.0.  Is there a different bug/issue open for Umbraco 8?

**Comment by nul800sebastiaan** [comment]
@bachicool This has only been fixed for v9. We're in support mode for v8 since the 8.18.0 release, which means we will fix regression and security issues on v8.

**Comment by m4nickroll** [comment]
@nul800sebastiaan This regressed in 8.14 so should arguably be fixed in v8 too? We've got a client with a media-heavy site and it's totally changed the behaviour (to be incorrect) when dragging and dropping images since upgrading to 8.17 from pre 8.14

**Comment by nul800sebastiaan** [comment]
I understand, we have many things that arguably should be fixed, but only so much time. At the moment we classify regressions as bugs introduced in the last 3 minors and unfortunately that makes 8.14 bugs no longer eligible.

**Comment by nul800sebastiaan** [comment]
@m4nickroll Understood! And of course we'll consider exceptions to the rule where they make sense.

However.. maybe I wasn't clear about this, this scenario exists in all of v7 and v8 with the default media picker as well, we've never adhered very well to what was allowed where. Consider the following `Custom Folder` that only allows media items of type `Custom Image` to be created under it in v8.11 - it just creates images of the original type "Image" without question, it also just creates a folder which is not allowed here:


https://user-images.githubusercontent.com/304656/158405386-54fe4c57-362e-4604-b97c-f59bcc6be9c3.mp4

**Comment by bachicool** [comment]
@nul800sebastiaan, Agree, that this bug does exist in the earlier versions. 

In the media section - With this bug, we can't drag and drop the images. The only alternative is to hide the drag and drop area using CSS. But if we hide the drag and drop functionality it will take away the end users/editor's best experience in using Umbraco. Meaning they will need to create the images one by one.  And this will restrict most of us with custom folders from upgrading Umbraco any higher than 8.13.  And I hope you appreciate the fact that on a website media and content are equally important. 

In my humble opinion, this bug needs to be addressed and is one of those exceptions, due to its disruptiveness in the end-user experience. 

Following is the video where Drag and drop working in version 8.12.1: 



https://user-images.githubusercontent.com/574356/159158657-262993a1-ed43-4c9e-bf35-73b2536c7964.mp4

**Comment by nul800sebastiaan** [comment]
@bachicool Fair enough, that is a different problem that I actually didn't look at very much, I was focused on the media picker instead. I didn't really look in the media section very much for this one.

For  now, you could solve this by using our event handlers and reject things that are not allowed. However, if anyone would like to try to port this PR to v8 then we'd be happy to have a look (target branch: `v8/dev`).

**Comment by nul800sebastiaan** [comment]
Honestly, I didn't realize there was a breaking change until two weeks ago when we talked about this, but I've marked it as a breaking change now referring to this PR. https://github.com/umbraco/Umbraco-CMS/pull/9461

**Comment by jgettrup-ng** [comment]
Thanks everyone for their work on this. After installing and testing v8.18.4 there is still an issue with this that we think represents a breaking change. We jumped from v8.12 to v8.18 so we don't know exactly when it changed.

We have a custom media type called CustomFolder that inherits from Folder and another called CustomImage that has an umbracoFile property. The permissions on CustomFolder allow just two child node types: CustomImage and CustomFolder (to allow for subfolder structures).

When we drag an image onto the upload control we get the disallowedMediaType message ("Cannot upload this file, the media type with alias 'Image' is not allowed here"). The expected behaviour is that a CustomImage item is created and the image file is uploaded into it.

The issue is that there are two allowed child node types (folder and image), so the code does not allow the automatic creation of the image item. I would argue that there is only one "valid" child node type (the one with the umbracoFile property), and that an item of that type should be created automatically. We believe this used to be the behaviour in v8.12.

I suggest one possible solution below, which checks for the number of allowed media types that have the umbracoFile property:

src/Umbraco.Web/Editors/MediaController.cs: line 711
if (parentId != Constants.System.Root)
{
    var mediaFolderItem = mediaService.GetById(parentId);
    var mediaFolderType = allMediaTypes.FirstOrDefault(x => x.Alias == mediaFolderItem.ContentType.Alias);
+   int numAllowedContentTypesWithFileProperty = 0;

    if (mediaFolderType != null)
    {
        IMediaType mediaTypeItem = null;

        foreach (ContentTypeSort allowedContentType in mediaFolderType.AllowedContentTypes)
        {
            IMediaType checkMediaTypeItem = allMediaTypes.FirstOrDefault(x => x.Id == allowedContentType.Id.Value);
            allowedContentTypes.Add(checkMediaTypeItem);

            var fileProperty = checkMediaTypeItem?.CompositionPropertyTypes.FirstOrDefault(x => x.Alias == Constants.Conventions.Media.File);
            if (fileProperty != null)
            {
                mediaTypeItem = checkMediaTypeItem;
 +              numAllowedContentTypesWithFileProperty++;
            }
        }

        //Only set the permission-based mediaType if we only allow 1 specific file under this parent.
 -      if (allowedContentTypes.Count == 1 && mediaTypeItem != null)
 +      if (numAllowedContentTypesWithFileProperty == 1 && mediaTypeItem != null)
        {
            mediaTypeAlias = mediaTypeItem.Alias;
        }
    }
}
else
{
    var typesAllowedAtRoot = allMediaTypes.Where(x => x.AllowedAsRoot).ToList();
    allowedContentTypes.UnionWith(typesAllowedAtRoot);
}

**Comment by jgettrup-ng** [comment]
It's a pity this one hasn't been fixed yet.

I did suggest to our editors that they use the "Create" button to create the media item first, and then drag the image file into the newly-created item, but they really missed the speed and simplicity of the drag and drop to create.

So, I ended up creating a new single-level "leaf" folder type that could only contain the image item and not the folder item. This allows the editors to drag and drop images again, but not create any further levels of subfolder. For our use case, the editors were happy with that workaround.


From: Chris Norwood ***@***.***>
Sent: 25 October 2022 17:58
To: umbraco/Umbraco-CMS ***@***.***>
Cc: Jim Gettrup ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [umbraco/Umbraco-CMS] (umbraCollab) Check media Parent for permissions when setting correct MediaType (PR #11858)

This Message originated outside your organization.
________________________________
We have also just come across this issue in Umbraco 10.3.1 (migrating from 8.6.5, v10 currently in UAT which is how we discovered this)

We have exactly the same structure as @jgettrup-ng<https://github.com/jgettrup-ng> - a custom folder with a custom media type where the custom media type is the only allowed media type in the folder (but the folder type is also allowed to allow for sub-folders). When dragging and dropping a file we get the error about not being allowed to create the "file" type.

-

**Pull request merged**